sphinx_gallery/recommender.py

```diff
@@ -0,0 +1,273 @@
+
+        Parameters
+        ----------
+        X : ndarray of shape (n_samples, n_features)
+            A term frequency matrix.
+
+        Returns
+        -------
+        X_tfidf : ndarray of shape (n_samples, n_features)
+            A tf-idf matrix of the same shape as X.
+        """
+        import numpy as np
+
+        n_samples, n_features = X.shape
+
+        df = np.count_nonzero(X, axis=0) + 1
+        n_samples += 1
+        idf = np.log(n_samples / df) + 1
+        X_tfidf = X * idf
+        X_tfidf = (X_tfidf.T / np.linalg.norm(X_tfidf, axis=1)).T
+
+        return X_tfidf
+
+    def cosine_similarity(self, X, Y=None):
+        """
+        Compute the cosine similarity between two vectors X and Y.
+
+        Parameters
+        ----------
+        X : ndarray of shape (n_samples_X, n_features)
+            Input data.
+
+        Y : ndarray of shape (n_samples_Y, n_features), default=None
+            Input data. If `None`, the output will be the pairwise
+            similarities between all samples in `X`.
+
+        Returns
+        -------
+        cosine_similarity : ndarray of shape (n_samples_X, n_samples_Y)
+            Cosine similarity matrix.
+        """
+        import numpy as np
+
+        if Y is X or Y is None:
+            Y = X
+
+        X_normalized = X / np.linalg.norm(X)
+        if X is Y:
+            Y_normalized = X_normalized
+        else:
+            Y_normalized = Y / np.linalg.norm(Y)
+        similarity = X_normalized @ Y_normalized.T
+
+        return similarity
+
+    def fit(self, file_names):
+        """
+        Compute the similarity matrix of a group of documents.
+
+        Parameters
+        ----------
+        file_names : list or generator of file names.
+
+        Returns
+        -------
+        self : object
+            Fitted recommender.
+        """
+        import numpy as np
+
+        n_examples = self.n_examples
+        min_df = self.min_df
+        max_df = self.max_df
+        if not isinstance(n_examples, numbers.Integral):
+            raise ValueError("n_examples must be an integer")
+        elif n_examples < 1:
+            raise ValueError("n_examples must be strictly positive")
+
+        freq_func = self.token_freqs
+        counts_matrix = self.dict_vectorizer(
+            [freq_func(Path(fname).read_text()) for fname in file_names]
+        )
+
+        if isinstance(min_df, float) and 0.0 < min_df <= 1.0:
+            min_df = int(np.ceil(min_df * counts_matrix.shape[0]))
+        elif not isinstance(min_df, numbers.Integral) and min_df > 1:
+            raise ValueError("min_df must be float in range [0.0, 1.0] or int")
+        if isinstance(max_df, float) and 0.0 < max_df <= 1.0:
+            max_df = int(np.floor(max_df * counts_matrix.shape[0]))
+        elif not isinstance(max_df, numbers.Integral) and max_df > 1:
+            raise ValueError("max_df must be float in range [0.0, 1.0] or int")
```

Comment:
> to keep validation out of `fit`

Sorry, I don't follow, validation of parameters? But this is currently in fit?

And sorry, I just meant earlier in `fit`, but now thinking about it, this could go in `gen_gallery` along with all the other config validation, as well? Let's us raise earlier if config is incorrect.

Reply:
Sorry, I bugged a little in my previous comment. I meant out of `__init__` and inside `fit`.

I prefer keeping the parameter validation inside the class to avoid polluting `gen_gallery`. Instead, I can rework the logic to carry the validation before vectorizing.

Reply:
Yeah config checking in `gen_gallery` is messy and I've been meaning to clean it up for a while, though it is nice to have all the checks in one place.

Reply:
Done in 6464c4fabd89ae09d2319dd4a43ab79a7b8fb52d.